modelerfour/modeler/modelerfour.ts

```diff
@@ -13,6 +13,51 @@ function is(value: any): asserts value is object | string | number | boolean {
   }
 }
 
+/**
+ * Contains usage information for one appearance of a schema.  Used to propagate
+ * context from the usage of a schema to all schemas it references.
+ */
+interface SchemaUsageDetails {
+  context?: SchemaContext,
+  knownMediaType?: string
+}
+
+function pushIfMissing<T>(targetArray: T[], newValue: T): void {
+  if (targetArray.indexOf(newValue) === -1) {
+    targetArray.push(newValue);
+  }
+}
+
+function trackSchemaUsage(schema: Schema, schemaUsage: SchemaUsageDetails): void {
+  const processedSchemas = new Set<Schema>();
+
+  function innerTrackSchemaUsage(schema: Schema) {
+    if (processedSchemas.has(schema)) {
+      return;
+    }
+
+    processedSchemas.add(schema);
+    if (schema instanceof ObjectSchema) {
+      if (schemaUsage.context) {
+        pushIfMissing(schema.usage = (schema.usage || []), schemaUsage.context);
+      }
+      if (schemaUsage.knownMediaType) {
+        pushIfMissing(schema.serializationFormats = (schema.serializationFormats || []), schemaUsage.knownMediaType);
+      }
+
+      schema.parents?.all?.forEach(p => innerTrackSchemaUsage(p));
```

Comment:
sweet mercy... the optional chaining operator sure makes life better.  _+1 hero points_

you missed the hero point for `schema.parents?.all?.forEach(innerTrackSchemaUsage)` 😆 

(not that it matters at all :D )

Reply:
This makes me wonder if there would be a way to go back and refactor all the places where the chaining operator would make code cleaner. *sigh*

Reply:
lol, you're right, I should have just dropped the function in directly :)  I wonder if there's an eslint code fix you could bulk-execute on the code?